### PR TITLE
udev: introduce OPTIONS="log_level=xxx" rule

### PR DESCRIPTION
**Pull request by yuwata**
cc @msekletar.

### CODE REVIEW

**Comment by msekletar** [comment]
Thanks a lot! 

I did very superficial review and changes LGTM.

File: src/udev/udev-rules.c [hunk 1]
```diff
@@ -1858,6 +1871,23 @@ static int udev_rule_apply_token_to_event(
...
+                        level = event->default_log_level;
+
+                log_set_max_level_realm(LOG_REALM_UDEV, level);
+                log_set_max_level_realm(LOG_REALM_SYSTEMD, level);
+
+                if (level == LOG_DEBUG && !event->log_level_was_debug) {
+                        /* The log level becomes LOG_DEBUG at first time. Let's log basic information. */
+                        log_device_uevent(dev, LOG_DEBUG, "Log level is changed to 'debug' while processing device");
+                        event->log_level_was_debug = true;
+                }
```
Comment (dnicolodi) [inline src/udev/udev-rules.c:1886]: I don't understand why the change in log level is reported only when switching to the DEBUG log level. Wouldn't it be less surprising if a change in log level would always be reported? It may also be worth to report the log level change for each event, not only the first time. I guess that the DEBUG log level is anyhow chatty enough that one more message is not going to matter much.

Comment (yuwata) [reply src/udev/udev-rules.c:1886]: > I don't understand why the change in log level is reported only when switching to the DEBUG log level. Wouldn't it be less surprising if a change in log level would always be reported? It may also be worth to report the log level change for each event, not only the first time. I guess that the DEBUG log level is anyhow chatty enough that one more message is not going to matter much.

The main motivation of this block is not reporting the log level change, but showing what uevent is currently processed. We usually show such information in DEBUG level. So, I do not think it is not necessary to log here in other level. When the main process is invoked with SYSTEMD_LOG_LEVEL=debug or something, then `worker_process_device()` logs the uevent information. However, if the log level of the main process is not DEBUG, then the information is not logged at the beginning of `worker_process_device()`, as at that time the maximum log level is not DEBUG. So, to obtain better debug logs, we should logs uevent info when the log level is changed to DEBUG in the first time.

File: src/shared/udev-util.c [hunk 2]
```diff
@@ -324,6 +324,26 @@ bool device_for_action(sd_device *dev, DeviceAction action) {
         return a == action;
 }
 
+int log_device_uevent(sd_device *device, int level, const char *str) {
```
Comment (poettering) [inline src/shared/udev-util.c]: our log functions typically don't fail (or if they fail we mostly ignore that). And I think that makes a lot of sense: logging should be mostly side-effect free, it should document code flow, not change code flow.

Hence maybe rework this to not propagate errors but instead to include any failures in the log message generated by this, so that we can just call this function, and it does the right thing and we can forget about the rest?

File: src/udev/udev-rules.c [hunk 3]
```diff
@@ -1858,6 +1871,23 @@ static int udev_rule_apply_token_to_event(
...
                 device_set_devlink_priority(dev, PTR_TO_INT(token->data));
                 break;
+        case TK_A_OPTIONS_LOG_LEVEL: {
+                int level = PTR_TO_INT(token->data);
+
+                if (level < 0)
+                        level = event->default_log_level;
+
+                log_set_max_level_realm(LOG_REALM_UDEV, level);
+                log_set_max_level_realm(LOG_REALM_SYSTEMD, level);
```
Comment (poettering) [inline src/udev/udev-rules.c]: maybe rework on top of #17914?

File: src/udev/udevd.c [hunk 4]
```diff
@@ -522,6 +523,10 @@ static int worker_device_monitor_handler(sd_device_monitor *monitor, sd_device *
         assert(dev);
         assert(manager);
 
+        /* Reset the log level, as it might be changed by "OPTIONS=log_level=". */
+        log_set_max_level_realm(LOG_REALM_UDEV, manager->log_level);
+        log_set_max_level_realm(LOG_REALM_SYSTEMD, manager->log_level);
```
Comment (poettering) [inline src/udev/udevd.c]: as above

**Review by poettering — APPROVED** [review]
lgtm

**Comment by yuwata** [comment]
@poettering Thank you for your review.
- Rebased on top ot #17914.
- Made `log_device_uevent()` not fail.

PTAL.

**Comment by boucman** [comment]
I'm not sure exactly how you envision the use of this feature but I think it might be worth adding a standard, late rule that unconditionnaly reset log_level, and maybe an early, never-active rule that enables it, to give people an idea where/how to use this feature ?
